src/main/groovy/com/google/protobuf/gradle/ProtobufExtension.groovy

```diff
@@ -53,7 +57,7 @@ abstract class ProtobufExtension {
    * The base directory of generated files. The default is
    * "${project.buildDir}/generated/source/proto".
    */
-  private String generatedFilesBaseDir
+  private final AtomicReference<String> generatedFilesBaseDir = new AtomicReference<String>()
```

Comment:
Why we need AtomicReference here?

Reply:
`ProtobufExtension` has a reference to `project`. I can't use project from the Action, as that is incompatible with configuration cache. So I moved the value of `generatedFilesBaseDir` to its own separate object so the Provider could reference that object without also referencing `project`.

Reply:
generatedFilesBaseDir is a String type. Where generatedFilesBaseDir has reference to project?

Reply:
Why we can't use `Provider<String>` instead of `AtomicReference<String>`?

Reply:
`Provider` is only for reading. I looked and didn't see any `Property<String>` impl (which confused me, but I also didn't care much). But I see now there is `ObjectFactory.property(Class)` that could be used. I don't know if that's any better though, since this isn't a Task, but I can swap to it.

Reply:
I think property is much more expected than java reference. Its better to use property. https://docs.gradle.org/current/userguide/lazy_configuration.html#connecting_properties_together

Reply:
And here https://docs.gradle.org/current/userguide/custom_plugins.html#sec:getting_input_from_the_build

Reply:
One more cool api feature. We can check if a user set the value [isPresent](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Provider.html#isPresent--) or not. Default value can be set with [convention](https://docs.gradle.org/current/javadoc/org/gradle/api/provider/Property.html#convention-T-)